devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy

```diff
@@ -0,0 +1,656 @@
+}
+
+private getAutoOnIntervalParam() {
+    return getParam(6, "Auto Turn-On Timer(0,Disabled; 1--60480 minutes)", 4, 0, null, "0..60480")
+}
+
+private getNightLightParam() {
+    return getParam(7, "Night Light Settings(1,10%;2,20%,...10,100%)", 1, 2, null, "1..10")
+}
+
+private getPowerFailureRecoveryParam() {
+    return getParam(8, "Power Failure Recovery", 1, 2, powerFailureRecoveryOptions)
+}
+
+private getPushDimmingDurationParam() {
+    return getParam(9, "Push Dimming Duration(0,Disabled; 1--10 Seconds)", 1, 2, null, "0..10")
+}
+
+private getHoldDimmingDurationParam() {
+    return getParam(10, "Hold Dimming Duration(1--10 Seconds)", 1, 4, null, "1..10")
+}
+
+private getMinimumBrightnessParam() {
+    return getParam(11, "Minimum Brightness(0,Disabled; 1--99:1%--99%)", 1, 10, null,"0..99")
+}
+
+private getMaximumBrightnessParam() {
+    return getParam(12, "Maximum Brightness(0,Disabled; 1--99:1%--99%)", 1, 99, null,"0..99")
+}
+
+private getTemperatureReportTimeParam() {
+    return getParam(13, "Temperature report time(1--60 minutes)", 1, 1, null, "1..60")
+}
+
+private getTemperatureReportThresholdParam() {
+    return getParam(14, "Temperature report threshold", 1, 5, temperatureReportThresholdOptions)
+}
+
+private getParam(num, name, size, defaultVal, options=null, range=null) {
+    def val = safeToInt((settings ? settings["configParam${num}"] : null), defaultVal)
+
+    def map = [num: num, name: name, size: size, value: val]
+    if (options) {
+        map.valueName = options?.find { k, v -> "${k}" == "${val}" }?.value
+        map.options = setDefaultOption(options, defaultVal)
+    }
+    if (range) map.range = range
+
+    return map
+}
+
+private static setDefaultOption(options, defaultVal) {
+    return options?.collectEntries { k, v ->
+        if ("${k}" == "${defaultVal}") {
+            v = "${v} [DEFAULT]"
+        }
+        ["$k": "$v"]
+    }
+}
+
+private static getLedModeOptions() {
+    return [
+            "0":"Off When On", "1":"On When On", "2":"Always Off", "3":"Always On"
+    ]
+}
+
+private static getPowerFailureRecoveryOptions() {
+    return [
+            "0":"Turn Off", "1":"Turn On", "2":"Restore Last State"
+    ]
+}
+
+private static getTemperatureReportThresholdOptions() {
+    def options = [:]
+    options["1"] = "1℃/1.8°F"
+    def it2
+
+    (2..10).each {
+        it2 = it*1.8
+        options["${it}"] = "${it}℃/${it2}°F"
+    }
+    return options
+}
+
+private sendEventIfNew(name, value, displayed=true, type=null, unit="") {
+    def desc = "${name} is ${value}${unit}"
+    if (device.currentValue(name) != value) {
+        logDebug(desc)
+
+        def evt = [name: name, value: value, descriptionText: "${device.displayName} ${desc}", displayed: displayed]
+
+        if (type) {
+            evt.type = type
+        }
+        if (unit) {
+            evt.unit = unit
+        }
+        sendEvent(evt)
+    }
+    else {
+        logTrace(desc)
+    }
+}
```

Comment:
This will be checked by Device API, remove this method and just send an event where you're calling it.

Reply:
Yes，you are right ！